Account/FirefoxAccount.swift

```diff
@@ -190,13 +196,104 @@ open class FirefoxAccount {
+        open let email: String
+        open let avatar: Avatar
+        
+        init(email: String, displayName: String?, avatar: String?) {
+            self.email = email
+            self.displayName = displayName
+            self.avatar = Avatar(url: avatar?.asURL)
+        }
+        
+        enum ImageDownloadState {
+            case notStarted
+            case started
+            case failedCanRetry
+            case failedCanNotRetry
+            case succeededMalformed
+            case succeeded
+        }
+        
+        open class Avatar {
+            open var image: UIImage?
+            open let url: URL?
+            var currentImageState: ImageDownloadState = ImageDownloadState.notStarted
+            
+            init(url: URL?) {
+                self.image = UIImage(named: "placeholder-avatar")
+                self.url = url
+                self.updateAvatarImageState()
+            }
+            
+            func updateAvatarImageState() {
+                switch currentImageState {
+                case .notStarted:
+                    self.currentImageState = ImageDownloadState.started
+                    self.downloadAvatar()
+                    break
+                case .failedCanRetry:
+                    self.downloadAvatar()
+                    break
+                default:
+                    break
+                }
+            }
+            
+            func downloadAvatar() {
+                SDWebImageManager.shared().downloadImage(with: url, options: [SDWebImageOptions.continueInBackground, SDWebImageOptions.lowPriority], progress: nil) { (image, error, cacheType, success, url) in
+                    if let error = error {
+                        if (error as NSError).code == 404 || self.currentImageState == ImageDownloadState.failedCanRetry {
+                            // Image is not found or failed to download a second time
+                            self.currentImageState = ImageDownloadState.failedCanNotRetry
+                        } else {
+                            // This could have been a transient error, attempt to download the image only once more
+                            self.currentImageState = ImageDownloadState.failedCanRetry
+                            self.updateAvatarImageState()
+                        }
+                        return
+                    }
+                    
+                    if success == true && image == nil {
+                        self.currentImageState = ImageDownloadState.succeededMalformed
+                        return
+                    }
+                    
+                    self.image = image
+                    self.currentImageState = ImageDownloadState.succeeded
+                    
+                    // Delay posting notification by 100ms to avoid any racing observers
+                    DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + .milliseconds(100)) {
```

Comment:
On which thread does `SDWebImage` run its callback? If the answer is 'main', then this really does just delay by 100msec. If not, then this is a change of behavior — the notification observers will run their code on the main thread.

Reply:
It appears to run the completion callback on `dispatch_main_sync_safe` for cached and downloaded images, [1](https://github.com/rs/SDWebImage/blob/edb4bc3554a8587d84d10ca930ba70166da430fa/SDWebImage/SDWebImageManager.m#L241), [2](https://github.com/rs/SDWebImage/blob/edb4bc3554a8587d84d10ca930ba70166da430fa/SDWebImage/SDWebImageManager.m#L229), [3](https://github.com/rs/SDWebImage/blob/edb4bc3554a8587d84d10ca930ba70166da430fa/SDWebImage/SDWebImageManager.m#L269).

Considering this and your earlier comment, should I lower the delay or remove it altogether?

Reply:
You should remove it and make sure that the observer is safe and performant to run on the main thread. If the notification is elsewhere sent from a non-main thread (which I believe it is) you should be prepared for the observers to run twice at the same time. If you don't want that to happen, you need to control which thread sends the notification.

Reply:
Tested this with several clients and it performed well when receiving the notifications.